### PR TITLE
Reportsystem

### PR DESCRIPTION
**Pull request by Allof02**
Implemented Report System 
newly added: ManipulateReviewDs, ManipulateUserDs 
Changed: User.java (add reportsystem associated attributes and methods) 
overall 80% done

### CODE REVIEW

File: src/main/java/report_use_cases/ReportInteractor.java
Comment (JarodGratton) [inline src/main/java/report_use_cases/ReportInteractor.java:31]: Should these be exceptions? I don't think this counts as exceptional circumstances

File: src/main/java/report_screens/ManipulateReviewDs.java
Comment (JarodGratton) [inline src/main/java/report_screens/ManipulateReviewDs.java:15]: This looks good, but I think this method should be moved into the ReviewGateway

**Review by JarodGratton — DISMISSED** [review]
Looks great, just two small things:

The method in ManipulateReviewDs should be moved into ReviewGateway so it can be with all the other methods pertaining to the review database.

In ReportInteractor, when a user has already reported a review it throws an exception. I believe we are only supposed to throw exceptions in exceptional circumstances, so this may not be appropriate. However, it works so let's not change it unless we have extra time.

File: src/main/java/report_use_cases/ReportInteractor.java
Comment (Allof02) [reply src/main/java/report_use_cases/ReportInteractor.java:31]: I think these should be. They are runtime exceptions; they will be raised when the user violates the requirements of sending a report (A user cannot be banned and identical report should not be sent again). The ReportScreen will catch those exceptions and show the error message (in terms of which specific violatoin is being made) to the user.

File: src/main/java/report_screens/ManipulateReviewDs.java
Comment (Allof02) [reply src/main/java/report_screens/ManipulateReviewDs.java:15]: Sure, do u want me to do that? Or you want to do that? I don't wanna mess up your ReviewGateway just in case lol

Comment (JarodGratton) [reply src/main/java/report_screens/ManipulateReviewDs.java:15]: Let's both do it so we hopefully do not have to resolve any conflicts at a later merge

Comment (Allof02) [reply src/main/java/report_screens/ManipulateReviewDs.java:15]: ok so should we wait until this pull request is merged?  then send a new pull request? does that work?

Comment (JarodGratton) [reply src/main/java/report_screens/ManipulateReviewDs.java:15]: Yeah I think that works

File: src/main/java/report_use_cases/DsModifier.java
Comment (prajogt) [inline src/main/java/report_use_cases/DsModifier.java:7]: Just to keep consistent should change object_to_update to camel case

File: src/main/java/report_use_cases/ReportInteractor.java
Comment (prajogt) [inline src/main/java/report_use_cases/ReportInteractor.java:59]: Not sure if I'm misunderstanding your process, but theres no presenter to show the success view that I can see
Let me know if it is implemented elsewhere

Comment (Allof02) [reply src/main/java/report_use_cases/ReportInteractor.java:62]: yeah no presenter is being used; all messages (success OR failure view with error message) are shown to user by a popout screen, JOptionPane.showMessageDialog, within the ReportScreen class

**Review by prajogt — APPROVED** [review]
Just some consistency notes, it looks like you used Python variable naming conventions, we need to decide if we want to refactor our project to that or stick with camel case across
I believe that for Reviews the banned user is handled in the screens layer (@JarodGratton correct me if i'm wrong), however it is handled in your interactor layer. We should decide on which we should do.

Otherwise looks like the functionality is there.

File: src/main/java/report_use_cases/ReportInteractor.java
Comment (JarodGratton) [reply src/main/java/report_use_cases/ReportInteractor.java:62]: I can't reply to @prajogt other comment so I'll just put it here: I was going to prevent banned users from doing anything at the screen layer, but we can check in both layers to make sure

Comment (prajogt) [reply src/main/java/report_use_cases/ReportInteractor.java:62]: Ah ok fair enough, make sure you change it in the CRC later then

Comment (JarodGratton) [reply src/main/java/report_use_cases/ReportInteractor.java:62]: Will do

Comment (Allof02) [reply src/main/java/report_use_cases/ReportInteractor.java:62]: Jarod i think i miss clicked something i dismissed ur approval; can u do it again? @JarodGratton

**Review by JarodGratton — APPROVED** [review]

**Pull request merged**
